Acá importamos las librerias que son necesarias para el resto del código. 

In [72]:
import pandas as pd
import numpy as np
import streamlit as st
import matplotlib.pyplot as plt

Invocamos los datos que hemos importado tales como las 'reservas'. 'reporte de ingreso año 2024' y el 'check-in' 

In [ ]:
df = pd.read_csv('reservas 27032024 27032025.csv', sep=';', skiprows=1, encoding='utf-8')
df2 = pd.read_csv('reporte de ingreso año 2024 a la fecha.csv', sep=';', encoding='latin1')
df3 = pd.read_excel('Check-in del 2024-03-27 al 2025-03-28 bookin.xls', sheet_name='Sheet1', header=0)

invocamos el DataFrame de 'df'

In [67]:
df

Titular                Tipo doc Número doc  \
0                   alba quintero  Cédula de ciudadanía  1067848216   
1                     jose acosta  Cédula de ciudadanía   111585061   
2                 jonathan VARGAS  Cédula de ciudadanía  1115913988   
3      Edgar Armando Mena Montoya                     NaN         NaN   
4                      juan gomez  Cédula de ciudadanía  1000019294   
...                           ...                     ...         ...   
2462         ana hernandez bustos  Cédula de ciudadanía   109876882   
2463                  A ldo Suros                     NaN         NaN   
2464             Daniel El Khoury               Pasaporte   23HE81140   
2465                 KARIM BOURAS               Pasaporte   20EA11230   
2466  John Stever Parra Rodriguez  Cédula de ciudadanía  1024514802   

     Nacionalidad Email Fecha de nacimiento    Género  \
0        Colombia   NaN          10/11/1985   Femenino   
1        Colombia   NaN                 NaN  Masculino   
2        Colombia   NaN           5/07/1992  Masculino   
3             NaN   NaN                 NaN        NaN   
4        Colombia   NaN          25/05/2002  Masculino   
...           ...   ...                 ...        ...   
2462     Colombia   NaN          27/04/1995   Femenino   
2463          NaN   NaN                 NaN  Masculino   
2464      Francia   NaN           3/06/1994  Masculino   
2465      Francia   NaN          24/12/1990  Masculino   
2466     Colombia   NaN           4/03/1991  Masculino   

                          Habitación          Ingreso           Salida  ...  \
0          TRIPLES PRIVADO SAN RAFAEL  26/03/2025 5:00  27/03/2025 5:00  ...   
1           4 o 5 PAX PRIVADO GUATAPE  26/03/2025 5:00  27/03/2025 5:00  ...   
2                   TITIRIBI TITIRIBI  26/03/2025 5:00  27/03/2025 5:00  ...   
3           LA CEJA DORMITORIO Ceja 6  27/03/2025 5:00  31/03/2025 5:00  ...   
4                   TITIRIBI TITIRIBI  25/03/2025 5:00  26/03/2025 5:00  ...   
...                               ...              ...              ...  ...   
2462  SANTA FE BAÑO EXTERNO SANTA FE  27/03/2024 5:00  31/03/2024 5:00  ...   
2463        LA CEJA DORMITORIO Ceja 3  27/03/2024 5:00  28/03/2024 5:00  ...   
2464         4 o 5 PAX PRIVADO PEÑOL  28/03/2024 5:00  29/03/2024 5:00  ...   
2465        LA CEJA DORMITORIO Ceja 5  28/03/2024 5:00  31/03/2024 5:00  ...   
2466        LA CEJA DORMITORIO Ceja 4  27/03/2024 5:00  28/03/2024 5:00  ...   

      Canal          Creada por   Fecha creación Código de la reserva  \
0       NaN             Jhon117   27/03/2025 1:12              15351926   
1       NaN             Jhon117  26/03/2025 22:40              15350047   
2       NaN             Jhon117  26/03/2025 22:40              15350041   
3       NaN             Jhon117  26/03/2025 15:21              15343929   
4       NaN             Jhon117  25/03/2025 22:14              15337114   
...     ...                 ...               ...                   ...   
2462    NaN  jaime andres rocha   28/03/2024 1:47              11398621   
2463    NaN  jaime andres rocha  27/03/2024 23:45              11397430   
2464    NaN  jaime andres rocha  27/03/2024 23:35              11397280   
2465    NaN  jaime andres rocha  27/03/2024 19:48              11394517   
2466    NaN  jaime andres rocha  27/03/2024 17:20              11392650   

     Total alojamiento        Estatus  ID Categoría  \
0              79900.0            NaN          32364   
1              80000.0            NaN          32367   
2              49900.0            NaN          32399   
3                  0.0  Cuenta pagada          32361   
4              50000.0            NaN          32399   
...                ...            ...            ...   
2462          520000.0            NaN          32398   
2463           32023.0  Cuenta pagada          32361   
2464          115572.0            NaN          32367   
2465           81000.0        

In [68]:
# Elimina los valores nulos para evitar errores
genero = df['Género'].dropna()

# Cuenta cuántos hay de cada género
conteo = genero.value_counts()

# Calcula el porcentaje
porcentaje = genero.value_counts(normalize=True) * 100

# Muestra resultados
print("Conteo por género:\n", conteo)
print("\nPorcentaje por género:\n", porcentaje)


Conteo por género:
 Género
Masculino    1336
Femenino      803
Name: count, dtype: int64

Porcentaje por género:
 Género
Masculino    62.459093
Femenino     37.540907
Name: proportion, dtype: float64


Los datos faltantes (missing_data) que hay en cada columna, osea los datos que la gente no registra. 

In [27]:
# ...existing code...

# Contar los datos faltantes en cada columna
missing_data = df.isnull().sum()
print(missing_data)

# ...existing code...

Titular                               0
Tipo doc                             85
Número doc                         103
Nacionalidad                        102
Email                              2118
Fecha de nacimiento                 180
Género                             328
Habitación                           0
Ingreso                               0
Salida                                0
Ocupación(Número de personas)       0
Plan                               2467
Canal                              2238
Creada por                            0
Fecha creación                       0
Código de la reserva                 0
Total alojamiento                     0
Estatus                            2065
ID Categoría                         0
Myallocator ID categoría          2467
Myallocator ID                     2467
Myallocator ID de agencia          2467
dtype: int64


En este código, buscamos eliminar los datos que no necesitamos cuyos datos faltantes exceden de los 2000 y a su vez los eliminamos de las columnas del DataFrame. 

In [28]:
# ...existing code...

# Contar los datos faltantes en cada columna
missing_data = df.isnull().sum()

# Definir el umbral de datos faltantes (más de 2000)
threshold = 2000

# Identificar las columnas que exceden el umbral
columns_to_drop = missing_data[missing_data > threshold].index

# Eliminar las columnas del DataFrame
df = df.drop(columns=columns_to_drop)

# Mostrar las columnas eliminadas
print(f"Columnas eliminadas: {list(columns_to_drop)}")

# ...existing code...

Columnas eliminadas: ['Email', 'Plan', 'Canal', 'Estatus', 'Myallocator ID categorÃ\xada', 'Myallocator ID', 'Myallocator ID de agencia']


La información de los datos que se han llenado en cada columna de los 2467 que se han registrado (entries)

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2467 entries, 0 to 2466
Data columns (total 15 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Titular                          2467 non-null   object 
 1   Tipo doc                         2382 non-null   object 
 2   Número doc                      2364 non-null   object 
 3   Nacionalidad                     2365 non-null   object 
 4   Fecha de nacimiento              2287 non-null   object 
 5   Género                          2139 non-null   object 
 6   Habitación                      2467 non-null   object 
 7   Ingreso                          2467 non-null   object 
 8   Salida                           2467 non-null   object 
 9   Ocupación(Número de personas)  2467 non-null   int64  
 10  Creada por                       2467 non-null   object 
 11  Fecha creación                  2467 non-null   object 
 12  Código de la reserv

In [45]:
df2

Fecha                      Cliente   Tipo doc  \
0      27/03/2025 15:48  Koneczna Magdalena koneczna        C.E   
1      27/03/2025 15:48  Koneczna Magdalena koneczna        C.E   
2      27/03/2025 10:46                Mateo Barrera        C.C   
3      26/03/2025 20:15                alba quintero        C.C   
4      26/03/2025 17:54                  jose acosta        C.C   
...                 ...                          ...        ...   
10047  27/03/2024 10:52                  Gerrit Lieb  Pasaporte   
10048  27/03/2024 10:14        Ivan Camilo Hernandez        C.C   
10049   27/03/2024 7:36                 Marthe dutey  Pasaporte   
10050   27/03/2024 0:27            Juan Roberto Soto  Pasaporte   
10051   27/03/2024 0:27            Juan Roberto Soto  Pasaporte   

             Producto       Valor           Forma de pago      Registrado por  \
0              CIFRUT    2,000.00                Efectivo             Jhon117   
1         Cargo extra   23,000.00                Efectivo             Jhon117   
2         Alojamiento   47,500.00  Transferencia bancaria             Jhon117   
3         Alojamiento   79,900.00                Efectivo             Jhon117   
4         Alojamiento   80,000.00  Transferencia bancaria             Jhon117   
...               ...         ...                     ...                 ...   
10047     Alojamiento  119,600.00                Efectivo  jaime andres rocha   
10048            AGUA    3,500.00                Efectivo  jaime andres rocha   
10049        POSTOBON    3,000.00                Efectivo  jaime andres rocha   
10050  CERVEZA PILSEN    6,000.00                Efectivo  jaime andres rocha   
10051        POSTOBON    3,000.00                Efectivo  jaime andres rocha   

        Reserva #  Transacción #  
0      15248106.0       12787798  
1      15248106.0       12787798  
2      15079614.0       12784271  
3      15351926.0       12780162  
4      15350047.0       12778441  
...           ...            ...  
10047  11351140.0        8750962  
10048  11386479.0        8750344  
10049         NaN        8748794  
10050  10450398.0        8748140  
10051  10450398.0        8748140  

[10052 rows x 9 columns]

Mostrar la informacion del segundo Data Frame "info()"

In [32]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10052 entries, 0 to 10051
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Fecha           10052 non-null  object 
 1   Cliente         8414 non-null   object 
 2   Tipo doc        8152 non-null   object 
 3   Producto        10052 non-null  object 
 4   Descripción     766 non-null    object 
 5   Valor           10052 non-null  object 
 6   Forma de pago   10052 non-null  object 
 7   Registrado por  10052 non-null  object 
 8   Reserva #       6722 non-null   float64
 9   Factura #       48 non-null     object 
 10  Transacción #   10052 non-null  int64  
 11  Comprobante #   3 non-null      float64
dtypes: float64(2), int64(1), object(9)
memory usage: 942.5+ KB


Contamos los datos que faltan en cada columna y lo mostramos. 

In [33]:
# ...existing code...

# Contar los datos faltantes en cada columna
missing_data = df2.isnull().sum()
print(missing_data)

# ...existing code...

Fecha                 0
Cliente            1638
Tipo doc           1900
Producto              0
Descripción        9286
Valor                 0
Forma de pago         0
Registrado por        0
Reserva #          3330
Factura #         10004
Transacción #         0
Comprobante #     10049
dtype: int64


En este codigo se procede a eliminar las columnas de los datos faltantes que se exceden mas de 9000 y se muestra abajo la lista de las columnas eliminadas.

In [44]:
# ...existing code...

# Contar los datos faltantes en cada columna de df2
missing_data_df2 = df2.isnull().sum()

# Definir el umbral de datos faltantes (más de 9000)
threshold_df2 = 9000

# Identificar las columnas que exceden el umbral
columns_to_drop_df2 = missing_data_df2[missing_data_df2 > threshold_df2].index

# Eliminar las columnas del DataFrame df2
df2 = df2.drop(columns=columns_to_drop_df2)

# Mostrar las columnas eliminadas
print(f"Columnas eliminadas de df2: {list(columns_to_drop_df2)}")

# ...existing code...

Columnas eliminadas de df2: ['Descripción', 'Factura #', 'Comprobante #']


In [73]:
data = {
    'Producto': ['CIFRUT', 'Cargo extra', 'CIFRUT', 'Cargo extra', 'CIFRUT'],
    'Cantidad': [1, 1, 1, 1, 1]
}
df = pd.DataFrame(data)

# Contar la cantidad de productos comprados
producto_counts = df['Producto'].value_counts()

# Calcular los porcentajes
producto_percentages = (producto_counts / producto_counts.sum()) * 100

# Crear la gráfica de barras
fig, ax = plt.subplots()
producto_counts.plot(kind='bar', ax=ax, color='skyblue', alpha=0.8)
ax.set_title('Cantidad de productos comprados')
ax.set_xlabel('Producto')
ax.set_ylabel('Cantidad')
ax.bar_label(ax.containers[0], labels=[f'{p:.1f}%' for p in producto_percentages])

# Mostrar la gráfica en Streamlit
st.pyplot(fig)

# Mostrar la tabla con cantidades y porcentajes
st.write("Tabla de productos comprados:")
st.write(pd.DataFrame({
    'Producto': producto_counts.index,
    'Cantidad': producto_counts.values,
    'Porcentaje': producto_percentages.values
}))

# ...existing code...

2025-04-14 16:20:14.403 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-04-14 16:20:14.797 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-04-14 16:20:15.205 
  command:

    streamlit run c:\Users\Yenifer\proyecto_1\santi\Lib\site-packages\ipykernel_launcher.py [ARGUMENTS]
2025-04-14 16:20:15.206 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-04-14 16:20:15.208 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-04-14 16:20:15.210 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-04-14 16:20:15.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored 

In [56]:
df3

Número de reserva              Reservado por  \
0            4396111470             Parra, John S.   
1            4557704876             López, Jojanna   
2            4637253976                Surós, Aldo   
3            4689813600             Le Roy, Victor   
4            4897263353      Hernandez, Ana Milena   
...                 ...                        ...   
2281         4502890135             moreno, liseth   
2282         4749774277  Trejos Cueto, José Manuel   
2283         4942788539        Magdalena, Koneczna   
2284         4517617109            Larouche, Mario   
2285         4540979937             Morales, Doris   

     Nombre del cliente (o clientes)     Entrada      Salida  \
0                      John S. Parra  2024-03-27  2024-03-28   
1                      Jojanna López  2024-03-27  2024-03-28   
2                         Aldo Surós  2024-03-27  2024-03-28   
3                      Victor Le Roy  2024-03-27  2024-03-28   
4               Ana Milena Hernandez  2024-03-27  2024-03-28   
...                              ...         ...         ...   
2281                   liseth moreno  2025-03-20  2025-03-23   
2282        José Manuel Trejos Cueto  2025-03-22  2025-03-26   
2283              Koneczna Magdalena  2025-03-23  2025-03-27   
2284                  Mario Larouche  2025-03-25  2025-03-27   
2285                   Doris Morales  2025-03-28  2025-04-04   

         Fecha de reserva              Estado  Habitaciones  Personas  \
0     2024-03-27 12:11:31                  ok             1         1   
1     2024-03-25 10:32:33  cancelled_by_guest             1         1   
2     2024-03-27 18:40:32                  ok             1         1   
3     2024-03-27 21:06:36                  ok             1         1   
4     2024-03-27 20:18:03                  ok             1         3   
...                   ...                 ...           ...       ...   
2281  2025-02-01 13:21:39  cancelled_by_guest             1         4   
2282  2025-01-22 08:00:41                  ok             1         1   
2283  2025-02-21 08:59:49                  ok             1         1   
2284  2025-02-13 18:07:20  cancelled_by_guest             1         1   
2285  2025-02-12 20:08:16  cancelled_by_guest             1         1   

          Precio  Comisión % Importe de la comisión  \
0      26910 COP          14             3767.4 COP   
1      26910 COP          14                    NaN   
2      26910 COP          14             3767.4 COP   
3      26910 COP          14             3767.4 COP   
4     107910 COP          14            15107.4 COP   
...          ...         ...                    ...   
2281  539730 COP          16                    NaN   
2282  162000 COP          16              25920 COP   
2283  110340 COP          16            17654.4 COP   
2284   42390 COP          16                    NaN   
2285  168975 COP          16                    NaN   

                                            Comentarios Booker country  \
0                                                   NaN             co   
1                                                   NaN             co   
2                                                   NaN             cr   
3                                                   NaN             co   
4                                                   NaN             co   
...                                                 ...            ...   
2281                                                NaN             co   
2282  Approximate time of arrival: between 10:00 and...             co   
2283  Approximate time of arrival: between 21:00 and...             pl   
2284                                                NaN             ca   
2285  Approximate time of arrival: between 15:00 and...             cl   

                                         Tipo de unidad  Duración (noches)  \
0                                            Dormitorio                  1   
1    

In [48]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2286 entries, 0 to 2285
Data columns (total 27 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Número de reserva                2286 non-null   int64  
 1   Reservado por                    2286 non-null   object 
 2   Nombre del cliente (o clientes)  2257 non-null   object 
 3   Entrada                          2286 non-null   object 
 4   Salida                           2286 non-null   object 
 5   Fecha de reserva                 2286 non-null   object 
 6   Estado                           2286 non-null   object 
 7   Habitaciones                     2286 non-null   int64  
 8   Personas                         2286 non-null   int64  
 9   Adultos                          28 non-null     float64
 10  Niños                            28 non-null     float64
 11  Edades de los niños:             1 non-null      float64
 12  Precio              

In [49]:
missing_data = df3.isnull().sum()
print(missing_data)

Número de reserva                     0
Reservado por                         0
Nombre del cliente (o clientes)      29
Entrada                               0
Salida                                0
Fecha de reserva                      0
Estado                                0
Habitaciones                          0
Personas                              0
Adultos                            2258
Niños                              2258
Edades de los niños:               2285
Precio                                0
Comisión %                            0
Importe de la comisión              890
Estado del pago                    2284
Forma de pago                      2286
Comentarios                        1717
Grupo de reserva                   2286
Booker country                        0
Motivo del viaje                    504
Dispositivo                           6
Tipo de unidad                        0
Duración (noches)                     0
Fecha de cancelación               1730


In [50]:
# ...existing code...

# Contar los datos faltantes en cada columna de df3
missing_data_df3 = df3.isnull().sum()

# Definir el umbral de datos faltantes (más de 2000)
threshold_df3 = 2000

# Identificar las columnas que exceden el umbral
columns_to_drop_df3 = missing_data_df3[missing_data_df3 > threshold_df3].index

# Eliminar las columnas del DataFrame df3
df3 = df3.drop(columns=columns_to_drop_df3)

# Mostrar las columnas eliminadas
print(f"Columnas eliminadas de df3: {list(columns_to_drop_df3)}")

# ...existing code...

Columnas eliminadas de df3: ['Adultos', 'Niños', 'Edades de los niños:', 'Estado del pago', 'Forma de pago', 'Grupo de reserva', 'Dirección', 'Número de teléfono']


In [52]:
# ...existing code...

# Contar la cantidad de registros por cada motivo del viaje
motivo_viaje_counts = df3['Motivo del viaje'].value_counts()
print(motivo_viaje_counts)

# ...existing code...

Motivo del viaje
Ocio       1574
Trabajo     208
Name: count, dtype: int64


In [53]:
# ...existing code...

# Eliminar la columna "Motivo del viaje" del DataFrame df3
df3 = df3.drop(columns=['Motivo del viaje'])

# Mostrar un mensaje confirmando la eliminación
print("La columna 'Motivo del viaje' ha sido eliminada.")

# ...existing code...

La columna 'Motivo del viaje' ha sido eliminada.


In [55]:
# ...existing code...

# Eliminar la columna "Dispositivo" del DataFrame df3
df3 = df3.drop(columns=['Dispositivo'])

# Mostrar un mensaje confirmando la eliminación
print("La columna 'Dispositivo' ha sido eliminada.")

La columna 'Dispositivo' ha sido eliminada.
